### PR TITLE
chore: run ci tests in parallel and get test insights

### PR DESCRIPTION
**Pull request by sspaink**
This PR adds parallelism to unit tests: https://circleci.com/docs/2.0/parallelism-faster-jobs/ we can adjust number of parallel jobs that get run to potentially further improve the time, currently all are set to 4.

Also adds the use of the project [gotestsum](https://github.com/gotestyourself/gotestsum) in order to get timings of each test. This adds a new insights into the tests when uploaded to circle-ci. More on test insights here: https://circleci.com/docs/2.0/insights/. The timing information is used by the CI tool to determine what tests to run in which jobs, to optimize the parallel runs.

When you click on one of the jobs that runs the unit tests (like `test-go-1_17`) you will now have four extra buttons linking you to the output of the separate jobs:

image

Here is a screenshot of the test insights showing number of tests and how long each takes.

image

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -48,7 +48,9 @@ endif
 
 # Go built-in race detector works only for 64 bits architectures.
 ifneq ($(GOARCH), 386)
-	race_detector := -race
+	ifneq ($(GOOS), windows)
```
[the remark this answers is not in the document]
Comment (sspaink) [reply Makefile:51]: Nice catch, must have been an historical reason `-race` wasn't been used with Windows before. I'll remove this extra check.

File: .circleci/config.yml [hunk 2]
```diff
@@ -25,6 +25,14 @@ executors:
       GOFLAGS: -p=8
 
 commands:
+  install-gotestsum:
+    parameters:
+      type:
+        type: string
+        default: "linux"
+    steps:
+      - run: curl -L https://github.com/gotestyourself/gotestsum/releases/download/v1.7.0/gotestsum_1.7.0_<< parameters.type >>_amd64.tar.gz --output gotestsum.tar.gz
```
[the remark this answers is not in the document]
Comment (sspaink) [reply .circleci/config.yml:34]: Moving the gotestsum version to one place sounds like a good idea to me, good suggestion.

**Comment by sspaink** [comment]
@powersj Thanks for the review! Took a bit more trail and error then I expected, but I've refactored the config to reduce duplicated versions/commands. Whenever you have sometime I'd appreciate another look at it, thanks!

summary of changes since last review:

- Moved windows/darwin test logic into the `test-go` command section, allows re-use of cmds & version
- Migrated `gotestsum` download script from powershell to bash so it can be used in all jobs and cached

**Review by powersj — APPROVED** [review]
Thanks for the clean-up! I'm sure finding all the differences from the OS X, Linux, and Windows runs was fun :)

Two comments on shell scripts:

1) Have you ever heard of or used [shellcheck](https://www.shellcheck.net/) before? I generally like to run those on any new shell scripts to ensure I have not missed anything. Minus some quoted variables, I think it looks good.

2) I always prefer to use long options (e.g. `--algorithm` vs `-a` or `--check` vs `-c`) to make it clearer what things are.

**Comment by sspaink** [comment]
Integrating shellcheck into our CI would probably be a good idea, created an issue for it: https://github.com/influxdata/telegraf/issues/9716

**Review by helenosheaa — APPROVED** [review]
looks great, good work!

**Pull request merged**